### PR TITLE
HPCC-21641 Cache GetComponentUsage report in ESP ws_machine

### PR DESCRIPTION
**Pull request by wangkx**
A user may bypass the cache function using BypassCachedResult option.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by wangkx** [comment]
@afishbeck please review.

File: esp/services/ws_machine/ws_machineService.hpp [hunk 1]
```diff
@@ -720,6 +721,144 @@ class CGetMachineInfoData
...
+
+    void add(const char* id, IArrayOf<IEspNodeGroupUsage>& usages)
+    {
+        CriticalBlock block(crit);
+        addUsageCacheElement(new MachineUsageCacheElement(id, usages));
+    }
+
+    std::list<Linked<MachineUsageCacheElement> > cache;
+    CriticalSection crit;
+    size32_t cacheSize;
```
Comment (afishbeck) [inline esp/services/ws_machine/ws_machineService.hpp]: Is this maxCacheSize?  Just calling it cacheSize is a little confusing.

**Review by afishbeck — COMMENTED** [review]
@wangkx looks good except comment, and does the cache get cleared on shutdown?  And have you run valgrind to make sure there are no leaks?

**Comment by wangkx** [comment]
@afishbeck The cache is cleared. I did run valgrind and it shows no leak. Please review my changes.

**Review by afishbeck — APPROVED** [review]
@wangkx looks good.

**Comment by richardkchapman** [comment]
@GordonSmith Does EclWatch UI need to be aware of the bypass cache option?

How were the figures of 10 seconds for the cache length and 8 for the cache size arrived at?

What is this information we are caching?

This all looks very specific code where I would have expected something a bit more generic to be available for general caching of EclWatch information?

**Comment by GordonSmith** [comment]
Yes - I would add a forced clear cache on manual refresh (user clicking the refresh button)
WRT the cache timeout, I would suggest an order of magnitude greater cache time (minutes not seconds):
* This service will get hit a lot as its now being called from the front page
* The results should be "slow changing" (free space on a hard drive).

**Comment by richardkchapman** [comment]
Back to @wangkx to address the timeout settings question.

**Comment by wangkx** [comment]
I chose 10 seconds because I thought the function is similar to Windows Task Manager which monitors a PC usage. But, this function here may take some time by collecting the usage from multiple machines. So, I am going to change the default timeout to 3 minutes. The cache can be bypassed by the bypass flag.

**Comment by wangkx** [comment]
@GordonSmith could you please review my changes?

**Review by GordonSmith — APPROVED** [review]
Code looks ok to me.

**Pull request merged**